src/libcore/ops/drop.rs

```diff
@@ -34,52 +34,139 @@
-/// `Outer`, then for `Inner`. Therefore, `main` prints `Dropping Outer!` and
-/// then `Dropping Inner!`.
+/// Here, `main` will print `Dropping!`. In other words, the compiler generates code that
+/// kind of looks like this:
+///
+/// ```rust,compile_fail,E0040
+/// # struct HasDrop;
+/// #
+/// # impl Drop for HasDrop {
+/// #     fn drop(&mut self) {
+/// #         println!("Dropping!");
+/// #     }
+/// # }
+/// fn main() {
+///     let _x = HasDrop;
 ///
+///     let mut x = _x;
+///     Drop::drop(&mut x);
+/// }
 /// ```
-/// struct Inner;
-/// struct Outer(Inner);
 ///
-/// impl Drop for Inner {
+/// Our custom implementation of `Drop` will be called at the end of `main`.
+///
+/// ## You cannot call `Drop::drop` yourself
+///
+/// Because the compiler automatically calls `Drop::drop`, you cannot call it yourself. This
+/// would lead to a "double drop", where `Drop::drop` is called twice on the same value. This
+/// can lead to things like "double frees".
+///
+/// In other words, if you tried to write the code in the previous example with the explicit
+/// call to `Drop::drop`, you'd get a compiler error.
+///
+/// If you'd like explicitly drop a value, [`std::mem::drop`] can be used instead.
+/// This function will take ownership of its argument and drop it. For more, see its documentation.
+///
+/// [`std::mem::drop`]: ../../std/mem/fn.drop.html
+///
+/// ## Destructors are recursive
+///
+/// If your type is something like a `struct` or `enum` that is an aggregate of other types,
+/// Rust will first call `Drop::drop` on the type, and then recursively on everything it contains.
+/// This happens even if the type does not implement `Drop` itself.
+/// For example:
+///
+/// ```rust
+/// struct HasDrop;
+///
+/// impl Drop for HasDrop {
 ///     fn drop(&mut self) {
-///         println!("Dropping Inner!");
+///         println!("Dropping HasDrop!");
 ///     }
 /// }
 ///
-/// impl Drop for Outer {
+/// struct HasTwoDrops {
+///     one: HasDrop,
+///     two: HasDrop,
+/// }
+///
+/// impl Drop for HasTwoDrops {
 ///     fn drop(&mut self) {
-///         println!("Dropping Outer!");
+///         println!("Dropping HasTwoDrops!");
 ///     }
 /// }
 ///
 /// fn main() {
-///     let _x = Outer(Inner);
+///     let _x = HasTwoDrops { one: HasDrop, two: HasDrop };
 /// }
 /// ```
 ///
-/// ## Variables are dropped in reverse order of declaration
+/// This will print
 ///
-/// `_first` is declared first and `_second` is declared second, so `main` will
-/// print `Declared second!` and then `Declared first!`.
+/// ```text
+/// Dropping HasTwoDrops!
+/// Dropping HasDrop!
+/// Dropping HasDrop!
+/// ```
+///
+/// Even if we remove the implementation of `Drop` for `HasTwoDrop`, the destructors of its fields
+/// are still called. This would result in
 ///
+/// ```test
+/// Dropping HasDrop!
+/// Dropping HasDrop!
 /// ```
-/// struct PrintOnDrop(&'static str);
 ///
-/// impl Drop for PrintOnDrop {
+/// ## Drop order
```

Comment:
IMO we can/should probably replace this with.

> The order in which values are dropped is always specified and can be viewed in **the reference**.

I believe that I did not care about the drop order as I started learning rust. So for me this would probably be the least relevant section.

Reply:
Many (if not most) C/C++ people care drop order.